### PR TITLE
Optimize seeding of levels by importing in batches

### PR DESCRIPTION
**Pull request by juanmanzojr**
Uses batch sizes when importing Levels in efforts to optimize the time it takes to `seed:custom_levels`.
Locally, 
Before using batches:
`Finished seed:custom_levels (35 minutes and 1 second)` 
After using batches:
`Finished seed:custom_levels (18 minutes and 30 seconds)`

### CODE REVIEW

File: dashboard/lib/level_loader.rb [hunk 1]
```diff
@@ -87,18 +87,17 @@ def self.import_levels(level_file_glob)
       immutable_level_columns = %i(id name created_at)
       update_columns = Level.columns.map(&:name).map(&:to_sym).
         reject {|column| immutable_level_columns.include? column}
-      Level.import! changed_levels, on_duplicate_key_update: update_columns
+      # Changed levels are too large to import them in batches larger than 100 due to serialized properties.
+      Level.import! changed_levels.sort_by(&:id), on_duplicate_key_update: update_columns, batch_size: 100
```
Comment (Hamms) [inline dashboard/lib/level_loader.rb:93]: What's the purpose of sorting the levels here?

File: dashboard/app/models/concerns/levels/levels_within_levels.rb [hunk 2]
```diff
@@ -73,6 +73,75 @@ def clone_child_levels(parent_level, new_suffix, editor_experiment: nil)
...
+      # contents of this level's `contained_level_names` property.
+      def setup_contained_levels_for(levels)
+        outdated_levels_child_level_ids = []
+        new_levels_contained_child_levels = []
+
+        levels.each do |level|
+          # if our existing contained levels already match the given names, do nothing
+          next if level.child_levels.contained.map(&:name) == level.contained_level_names
+
+          outdated_levels_child_level_ids += level.levels_child_levels.contained.ids
```
Comment (Hamms) [inline dashboard/app/models/concerns/levels/levels_within_levels.rb]: It occurs to me that there's probably potential for further optimization here; rather than deleting and recreating all relationships when we detect any mismatch, we could theoretically identify the subset of relationships that actually need updating, and only make the necessary changes.

Probably outside the scope of this PR, but perhaps a good follow-up!

File: dashboard/app/models/concerns/levels/levels_within_levels.rb [hunk 3]
```diff
@@ -73,6 +73,75 @@ def clone_child_levels(parent_level, new_suffix, editor_experiment: nil)
...
+          ParentLevelsChildLevel.import!(new_levels_contained_child_levels, batch_size: 1000)
+        end
+
+        levels.each do |level|
+          level.levels_child_levels.reset
+          level.child_levels.reset
+        end
+      end
+
+      def setup_project_template_level_for(levels)
```
Comment (Hamms) [inline dashboard/app/models/concerns/levels/levels_within_levels.rb]: I'd love to see a comment for this method like there is for `setup_contained_levels_for`!

File: dashboard/app/models/concerns/levels/levels_within_levels.rb [hunk 4]
```diff
@@ -73,6 +73,75 @@ def clone_child_levels(parent_level, new_suffix, editor_experiment: nil)
 
         return update_params
       end
+
+      # Create ParentLevelsChildLevel many-to-many relationships based on the
+      # contents of this level's `contained_level_names` property.
```
Comment (Hamms) [inline dashboard/app/models/concerns/levels/levels_within_levels.rb]: nit; 
```suggestion
      # contents of the specified levels' `contained_level_names` properties.
```

File: dashboard/test/models/concerns/levels/levels_within_levels_test.rb [hunk 5]
```diff
@@ -116,7 +116,7 @@ class LevelsWithinLevelsTest < ActiveSupport::TestCase
     # can reorder
     second_contained = create :free_response
     level.update!(contained_level_names: [first_contained.name, second_contained.name])
-    assert_equal [first_contained, second_contained], level.child_levels.contained
+    assert_equal [first_contained, second_contained], level.child_levels.contained.to_a
```
Comment (Hamms) [inline dashboard/test/models/concerns/levels/levels_within_levels_test.rb]: why is this necessary here, but not on line 121 below?

**Review by Hamms — COMMENTED** [review]
This looks great! Batching makes a ton of sense for this use case, and I love how this implementation preserves the existing methods and generally makes things a little more clear. I have a few nits and questions, but no major concerns.

I do have one medium-sized concern, however; I don't love how much repeated logic there is between `setup_contained_levels_for` and `setup_project_template_level_for`. The current batchless implementation was already a bit less DRY than I'd prefer, and with the new batching logic in place it's even more so.

What would you think about combining these two methods into one? Something like `setup_child_levels_for` which will set up both flavors of child levels; contained and project templates. It seems like everywhere we want to do one of the two we currently also want to do the other, but we could add an optional parameter or something to only consider a particular subset if we needed it.

**Comment by juanmanzojr** [comment]
> This looks great! Batching makes a ton of sense for this use case, and I love how this implementation preserves the existing methods and generally makes things a little more clear. I have a few nits and questions, but no major concerns.
> 
> I do have one medium-sized concern, however; I don't love how much repeated logic there is between `setup_contained_levels_for` and `setup_project_template_level_for`. The current batchless implementation was already a bit less DRY than I'd prefer, and with the new batching logic in place it's even more so.
> 
> What would you think about combining these two methods into one? Something like `setup_child_levels_for` which will set up both flavors of child levels; contained and project templates. It seems like everywhere we want to do one of the two we currently also want to do the other, but we could add an optional parameter or something to only consider a particular subset if we needed it.

That sounds like a good idea! lll get working on that

File: dashboard/lib/level_loader.rb (hunk 1 above)
Comment (artem-vavilov) [reply dashboard/lib/level_loader.rb:91]: https://dev.mysql.com/doc/refman/8.0/en/optimizing-innodb-bulk-data-loading.html#:~:text=When%20performing%20bulk%20inserts%2C%20it%20is%20faster%20to%20insert%20rows%20in%20PRIMARY%20KEY%20order

Comment (artem-vavilov) [reply dashboard/lib/level_loader.rb:93]: [10.5.5 Bulk Data Loading for InnoDB Tables: When performing bulk inserts, it is faster to insert rows in PRIMARY KEY order](https://dev.mysql.com/doc/refman/8.0/en/optimizing-innodb-bulk-data-loading.html#:~:text=When%20performing%20bulk%20inserts%2C%20it%20is%20faster%20to%20insert%20rows%20in%20PRIMARY%20KEY%20order)

Comment (Hamms) [reply dashboard/lib/level_loader.rb:93]: Perfect, thanks! I'd love to see a comment with that link, or at least one that references this implementation detail.

File: dashboard/app/models/concerns/levels/levels_within_levels.rb [hunk 6]
```diff
@@ -73,6 +73,50 @@ def clone_child_levels(parent_level, new_suffix, editor_experiment: nil)
...
+          # Skip if the current level's child levels already match the names
+          next if level.child_levels.send(level_kind).map(&:name) == level_names
+
+          # Collect outdated child levels' ids
+          outdated_levels_child_level_ids += level.levels_child_levels.send(level_kind).ids
+
+          next if level_names.blank?
+
+          # Create new relationships for matching names
+          Level.where(name: level_names).find_each do |contained_level|
```
Comment (Hamms) [inline dashboard/app/models/concerns/levels/levels_within_levels.rb]: nit:
```suggestion
          Level.where(name: level_names).find_each do |child_level|
```

File: dashboard/app/models/concerns/levels/levels_within_levels.rb [hunk 7]
```diff
@@ -73,6 +73,50 @@ def clone_child_levels(parent_level, new_suffix, editor_experiment: nil)
 
         return update_params
       end
+
+      # Create ParentLevelsChildLevel many-to-many relationships based on the
+      # contents of the specified levels' `contained_level_names` or `project_template_level_name` properties.
+      # Differentiated by level_kind (either contained or project template)
+      def setup_child_levels_for(levels, level_kind)
```
Comment (Hamms) [inline dashboard/app/models/concerns/levels/levels_within_levels.rb]: I'd love to be more consistent about making the distinction between parent levels and child levels clear throughout this method; in particular, I think it would be helpful to always make sure we refer to a child level as `child_level` rather than just `level`; ie, 

```suggestion
      def setup_child_levels_for(levels, child_level_kind)
```

File: dashboard/app/models/concerns/levels/levels_within_levels.rb [hunk 8]
```diff
@@ -73,6 +73,50 @@ def clone_child_levels(parent_level, new_suffix, editor_experiment: nil)
...
+        outdated_levels_child_level_ids = []
+        new_levels_child_levels = []
+
+        levels.each do |level|
+          # Determine which level names to use based on level_kind
+          level_names = level_kind == ParentLevelsChildLevel::CONTAINED ? level.contained_level_names : [level.project_template_level_name]
```
Comment (Hamms) [inline dashboard/app/models/concerns/levels/levels_within_levels.rb]: Same as above, I'd love to do something like this instead:
```suggestion
          child_level_names = child_level_kind == ParentLevelsChildLevel::CONTAINED ? level.contained_level_names : [level.project_template_level_name]
```

File: dashboard/app/models/concerns/levels/levels_within_levels.rb [hunk 9]
```diff
@@ -73,6 +73,50 @@ def clone_child_levels(parent_level, new_suffix, editor_experiment: nil)
...
+
+          next if level_names.blank?
+
+          # Create new relationships for matching names
+          Level.where(name: level_names).find_each do |contained_level|
+            new_levels_child_levels << ParentLevelsChildLevel.new(
+              parent_level: level,
+              child_level: contained_level,
+              kind: level_kind,
+              position: level_names.index(contained_level.name)
```
Comment (Hamms) [inline dashboard/app/models/concerns/levels/levels_within_levels.rb]: Project template levels don't need a position value. I doubt it would actively _hurt_ anything to include it, but it is a bit messy and I worry about the potential for confusion. Could we update this to only specify `position` for contained levels?

File: dashboard/app/models/concerns/levels/levels_within_levels.rb [hunk 10]
```diff
@@ -73,6 +73,50 @@ def clone_child_levels(parent_level, new_suffix, editor_experiment: nil)
 
         return update_params
       end
+
+      # Create ParentLevelsChildLevel many-to-many relationships based on the
+      # contents of the specified levels' `contained_level_names` or `project_template_level_name` properties.
+      # Differentiated by level_kind (either contained or project template)
```
Comment (Hamms) [inline dashboard/app/models/concerns/levels/levels_within_levels.rb]: nit; for clarity, please specify

```suggestion
      # Differentiated by level_kind (either ParentLevelsChildLevel::CONTAINED or ParentLevelsChildLevel::PROJECT_TEMPLATE)
```

**Review by Hamms — COMMENTED** [review]
This looks great! I love the new combined method, and appreciate how clean in general the implementation is. I have a few nitpicks about comments and variable naming that I'd like to see, but nothing critical.

Here's the speedup I see on my local machine on the latest commit:
With optimization: `Finished seed:custom_levels (8 minutes and 30 seconds)`
Without optimization: `Finished seed:custom_levels (13 minutes and 34 seconds)`

**Review by Hamms — APPROVED** [review]
Looks good to me! I have neither questions nor nitpicks; this is clear, readable, and looks like it should result in a healthy time savings for a huge number of automated builds. Thank you! :tada:

**Comment by artem-vavilov** [comment]
> it should result in a healthy time savings for a huge number of automated builds

Additionally, we could further optimize the `Level.setup_child_levels_for` steps by preloading all the necessary data, potentially achieving an additional 25% or more reduction in execution time.

**Comment by Hamms** [comment]
I'd love to see that as a follow-up! Would be curious to hear more about what data specifically you're thinking about

**Comment by artem-vavilov** [comment]
> I'd love to see that as a follow-up! Would be curious to hear more about what data specifically you're thinking about

https://github.com/code-dot-org/code-dot-org/blob/5f5c9ab7671679eaac4bb5f88d81b6a9d892432c/dashboard/app/models/concerns/levels/levels_within_levels.rb#L40-L43
https://github.com/code-dot-org/code-dot-org/blob/5f5c9ab7671679eaac4bb5f88d81b6a9d892432c/dashboard/lib/level_loader.rb#L86-L91
https://github.com/code-dot-org/code-dot-org/blob/5f5c9ab7671679eaac4bb5f88d81b6a9d892432c/dashboard/app/models/concerns/levels/levels_within_levels.rb#L91-L97
